File: Lab8/C#/Lab8.2/Car.cs
using System;

namespace CarLibrary
{
    public class Car
    {
        private readonly double Acceleration, SpeedLimit;
        private double Speed;
        private bool IsRiding;

        public Car(double acceleration, double speedLimit)
        {
            Acceleration = acceleration;
            SpeedLimit = speedLimit;
        }

        public void Ride()
        {
            if (!IsRiding)
            {
                IsRiding = true;
                StartRiding(this, new EventArgs());
            }
        }

        public void Tick()
        {
            if (IsRiding)
            {
                Speed += Acceleration;
                if (Speed > SpeedLimit)
                {
                    SpeedLimitExceeded(this, new SpeedLimitExceededEventArgs(Speed));
                }
            }
        }

        public void Stop()
        {
            if (IsRiding)
            {
                IsRiding = false;
                Speed = 0;
                StopRiding(this, new EventArgs());
            }
        }

        public EventHandler StartRiding;

        public EventHandler StopRiding;

        public EventHandler<SpeedLimitExceededEventArgs> SpeedLimitExceeded;
    }

    public class SpeedLimitExceededEventArgs : EventArgs
    {
        public readonly double Speed;

        public SpeedLimitExceededEventArgs(double speed)
        {
            Speed = speed;
        }
    }
}

File: Lab8/C++/Lab8.2/Lab8.2/Car.cpp
//
//  car.cpp
//  Lab8.2
//
//  Created by Maxim Trokhimets on 04.05.2020.
//  Copyright © 2020 Maxim Trokhimets. All rights reserved.
//

#include <stdio.h>
#include "Car.h"

namespace CarLibrary
{

    Car::Car(double acceleration, double speedLimit) : Acceleration(acceleration), SpeedLimit(speedLimit)
    {
    }

    void Car::Ride()
    {
        if (!IsRiding)
        {
            IsRiding = true;
            EventArgs tempVar();
            StartRiding(this, &tempVar);
        }
    }

    void Car::Tick()
    {
        if (IsRiding)
        {
            Speed += Acceleration;
            if (Speed > SpeedLimit)
            {
                SpeedLimitExceededEventArgs tempVar(Speed);
                SpeedLimitExceeded(this, &tempVar);
            }
        }
    }

    void Car::Stop()
    {
        if (IsRiding)
        {
            IsRiding = false;
            Speed = 0;
            EventArgs tempVar();
            StopRiding(this, &tempVar);
        }
    }

    SpeedLimitExceededEventArgs::SpeedLimitExceededEventArgs(double speed) : Speed(speed)
    {
    }
}
